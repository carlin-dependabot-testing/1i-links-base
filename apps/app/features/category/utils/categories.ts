import { DEFAULT_CATEGORY } from '@/constants'
import { type Category } from '@/features/links/types/category.types'
import { getCategories } from '@/features/setup/utils/load-json-data'

/**
 * Returns a Set of category names in lowercase
 */
export const getCategoryNames = async (): Promise<Set<string>> => {
  const categories = await getCategories()

  return new Set([
    DEFAULT_CATEGORY,
    ...(categories?.map(cat => cat.name.toLowerCase()) || [])
  ])
}

/**
 * Returns all valid categories
 */
export const getValidCategories = async (): Promise<Category[]> => {
  const categories = await getCategories()
  return categories || []
}

/**
 * Validates if a category name exists
 */
export const isCategoryValid = async (
  categoryName: string
): Promise<boolean> => {
  const categoryNames = await getCategoryNames()
  return categoryNames.has(categoryName.toLowerCase())
}

/**
 * Returns a category by its slug
 */
export const getCategoryBySlug = async (slug: string) => {
  const categories = await getCategories()
  return categories?.find(category => category.name === slug) || null
}

// Synchronous versions with default values
export const getDefaultCategoryNames = (): Set<string> => {
  return new Set([DEFAULT_CATEGORY])
}

export const getDefaultCategories = (): Category[] => {
  return []
}
