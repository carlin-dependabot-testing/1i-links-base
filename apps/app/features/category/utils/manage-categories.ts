import fs from 'fs/promises'
import path from 'path'
import { z } from 'zod'

import { CategorySchema } from '@/features/links/schemas/category.schema'
import { type Category } from '@/features/links/types/category.types'

// Custom errors
export class CategoryError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'CategoryError'
  }
}

export const categoriesPath = path.join(
  process.cwd(),
  'public',
  'data',
  'categories.json'
)

/**
 * Retrieves all categories from the JSON file
 * @returns {Promise<Category[]>} Array of categories
 * @throws {CategoryError} If there's an error reading the file
 */
export async function getCategories(): Promise<Category[]> {
  try {
    const data = await fs.readFile(categoriesPath, 'utf8')

    if (data.trim().length === 0) {
      console.warn('The category file is empty')
      return []
    }

    try {
      const parsedData = JSON.parse(data)

      // Sort categories alphabetically by name (case-insensitive)
      return parsedData.sort((a: Category, b: Category) =>
        a.name.toLowerCase().localeCompare(b.name.toLowerCase())
      )
    } catch (parseError) {
      console.error('Error parsing JSON:', parseError)
      return []
    }
  } catch (error) {
    if (error instanceof z.ZodError) {
      console.error('Zod validation error:', error.errors)
      throw new CategoryError('Invalid data format in categories file')
    }
    console.error('Error reading category:', error)
    return []
  }
}

/**
 * Saves categories to the JSON file
 * @param {Category[]} category - Array of categories to save
 * @throws {CategoryError} If there's an error saving the file
 */
export async function saveCategories(category: Category[]): Promise<void> {
  try {
    const data = JSON.stringify(category, null, 2)
    await fs.writeFile(categoriesPath, data)
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new CategoryError('Invalid category data format')
    }
    throw new CategoryError('Failed to save categories')
  }
}

/**
 * Adds a new category to the storage
 * @param {Category} newCategory - Category to add
 * @returns {Promise<Category>} Added category
 * @throws {CategoryError} If the category is invalid or if there's an error saving
 */
export async function addCategory(category: Category): Promise<Category> {
  try {
    CategorySchema.parse(category)

    const categories = await getCategories()

    categories.push(category)
    await saveCategories(categories)

    return category
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new CategoryError('Invalid category data format')
    }
    throw new CategoryError('Failed to add category')
  }
}

export async function editCategory(
  oldName: string,
  newName: string,
  icon?: string,
  description?: string
): Promise<void> {
  try {
    if (!oldName?.trim() || !newName?.trim()) {
      throw new CategoryError('Invalid category name')
    }

    const categories = await getCategories()

    const categoryExists = categories.some(
      category => category.name === oldName
    )

    if (!categoryExists) {
      throw new CategoryError('Category not found')
    }

    const isDuplicate = categories.some(
      category => category.name === newName && category.name !== oldName
    )
    if (isDuplicate) {
      throw new CategoryError('A category with this name already exists')
    }

    const updatedCategories = categories.map((category: Category) => {
      if (category.name === oldName) {
        const updatedCategory = {
          ...category,
          name: newName,
          icon: icon || category.icon,
          description: description || category.description
        }

        // Validate the updated category
        CategorySchema.parse(updatedCategory)
        return updatedCategory
      }
      return category
    })

    await saveCategories(updatedCategories)
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new CategoryError('Invalid category data format')
    }
    throw new CategoryError('Failed to edit category')
  }
}

/**
 * Deletes a category from the storage
 * @param {string} name - Name of the category to delete
 * @throws {CategoryError} If the category name is invalid
 */
export async function deleteCategory(name: string): Promise<void> {
  try {
    if (typeof name !== 'string' || name.trim() === '') {
      throw new CategoryError('Invalid category name')
    }

    const categories = await getCategories()
    const updatedCategories = categories.filter(
      category => category.name !== name
    )
    await saveCategories(updatedCategories)
  } catch (error) {
    if (error instanceof CategoryError) {
      throw new CategoryError('Failed to delete category')
    }
    throw error
  }
}
