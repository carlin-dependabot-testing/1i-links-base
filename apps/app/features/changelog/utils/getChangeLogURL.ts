import appConfig from '@/config/app.config'

export function getChangeLogURL() {
  return `https://raw.githubusercontent.com/${appConfig.githubRepo}/refs/heads/main/CHANGELOG.md`
}
