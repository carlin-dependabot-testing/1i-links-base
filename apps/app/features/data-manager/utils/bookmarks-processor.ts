import { DEFAULT_CATEGORY, DEFAULT_COLOR } from '@/constants'
import { type LinksApp } from '@/features/links/types/link.types'

export async function processBookmarksFile(
  file: File
): Promise<Partial<LinksApp>[]> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()

    reader.onload = event => {
      try {
        const text = event.target?.result as string
        const parser = new DOMParser()
        const doc = parser.parseFromString(text, 'text/html')
        const bookmarks = Array.from(doc.getElementsByTagName('a'))

        const links = bookmarks.map(bookmark => ({
          name: bookmark.textContent || '',
          environments: {
            production: bookmark.href
          },
          category: DEFAULT_CATEGORY,
          color: DEFAULT_COLOR
        }))

        resolve(links.filter(link => link.name && link.environments.production))
      } catch (error) {
        reject(new Error('Failed to process bookmarks file'))
      }
    }

    reader.onerror = () => {
      reject(new Error('Failed to read bookmarks file'))
    }

    reader.readAsText(file)
  })
}
