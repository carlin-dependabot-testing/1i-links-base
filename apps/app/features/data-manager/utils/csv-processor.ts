import { type LinksApp } from '@/features/links/types/link.types'

export async function processCSVFile(file: File): Promise<Partial<LinksApp>[]> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()

    reader.onload = event => {
      try {
        const text = event.target?.result as string
        const lines = text.split('\n')
        const headers = lines[0]
          .toLowerCase()
          .split(',')
          .map(header => header.trim())

        const links = lines
          .slice(1)
          .filter(line => line.trim()) // Skip empty lines
          .map(line => {
            const values = line.split(',').map(value => {
              // Remove quotes and trim
              return value.trim().replace(/^"(.*)"$/, '$1')
            })

            const link: Partial<LinksApp> = {
              environments: { production: '' }
            }

            headers.forEach((header, index) => {
              const value = values[index]

              if (!value) return

              switch (header) {
                case 'name':
                  link.name = value
                  break
                case 'description':
                  link.description = value
                  break
                case 'category':
                  link.category = value
                  break
                case 'color':
                  link.color = value
                  break
                case 'tags':
                  link.tags = value
                    .split(';')
                    .map(tag => tag.trim())
                    .filter(Boolean)
                  break
                case 'production_url':
                  if (!link.environments) link.environments = { production: '' }
                  link.environments.production = value
                  break
                case 'staging_url':
                  if (!link.environments) link.environments = { production: '' }
                  link.environments.staging = value
                  break
                case 'integration_url':
                  if (!link.environments) link.environments = { production: '' }
                  link.environments.integration = value
                  break
              }
            })

            return link
          })
          .filter(link => {
            // Filter out invalid links
            return link.name && link.environments?.production
          })

        resolve(links)
      } catch (error) {
        console.error('CSV Processing Error:', error)
        reject(
          new Error('Failed to process CSV file. Please check the file format.')
        )
      }
    }

    reader.onerror = () => {
      reject(new Error('Failed to read CSV file'))
    }

    reader.readAsText(file)
  })
}
