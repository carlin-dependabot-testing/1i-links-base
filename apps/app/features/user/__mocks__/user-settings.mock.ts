import { type UserSettings } from '@/features/user/stores/useUserSettingsStore'

export const mockUserSettings: UserSettings = {
  cardSize: 'default',
  cardSizeConfigs: {
    compact: {
      containerClasses: '',
      imageClasses: '',
      titleClasses: '',
      padding: 'p-2'
    },
    default: {
      containerClasses: '',
      imageClasses: '',
      titleClasses: '',
      padding: 'p-4'
    }
  },
  showGradients: true,
  showFavorites: false,
  hiddenCategories: [],
  newLinksInHiddenCategories: {},
  setHiddenCategories: jest.fn(),
  toggleHiddenCategory: jest.fn(),
  setCardSize: jest.fn(),
  setShowGradients: jest.fn(),
  setShowFavorites: jest.fn(),
  showDescriptions: true,
  showTags: true,
  showIcons: true,
  showEnvironmentBadges: true,
  setShowDescriptions: jest.fn(),
  setShowTags: jest.fn(),
  setShowIcons: jest.fn(),
  setShowEnvironmentBadges: jest.fn(),
  updateNewLinksCount: jest.fn(),
  resetNewLinksCount: jest.fn(),
  resetToDefaults: jest.fn(),
  previousDescriptionState: true,
  previousTagsState: true
} as const
