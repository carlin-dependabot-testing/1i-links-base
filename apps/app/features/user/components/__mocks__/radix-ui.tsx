import React from 'react'

// Mock Radix UI's RadioGroup components
jest.mock('@radix-ui/react-radio-group', () => ({
  Root: ({ children, onValueChange, value }: any) => (
    <div
      data-testid="radio-group"
      onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
        onValueChange?.(e.target.value)
      }
      data-value={value}
    >
      {children}
    </div>
  ),
  Item: ({ value, id }: any) => (
    <input
      type="radio"
      data-testid="radio-item"
      value={value}
      id={id}
      onChange={jest.fn()}
    />
  )
}))

// Mock Radix UI's Tooltip components
jest.mock('@radix-ui/react-tooltip', () => ({
  Root: ({ children }: any) => <div data-testid="tooltip-root">{children}</div>,
  Trigger: ({ children }: any) => (
    <div data-testid="tooltip-trigger">{children}</div>
  ),
  Content: ({ children }: any) => (
    <div data-testid="tooltip-content">{children}</div>
  ),
  Provider: ({ children }: any) => (
    <div data-testid="tooltip-provider">{children}</div>
  )
}))
