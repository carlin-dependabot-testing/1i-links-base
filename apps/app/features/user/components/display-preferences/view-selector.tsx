'use client'

import { Label } from '@links-base/ui/label'
import { RadioGroup, RadioGroupItem } from '@links-base/ui/radio-group'
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger
} from '@links-base/ui/tooltip'
import { cn } from '@links-base/ui/utils'

import { type CardSize } from '../../stores/useUserSettingsStore'

type ViewSelectorProps = {
  cardSize: CardSize
  onViewChange: (value: CardSize) => void
}

export const ViewSelector = ({ cardSize, onViewChange }: ViewSelectorProps) => {
  return (
    <div className="grid gap-2">
      <Label>View</Label>
      <RadioGroup
        value={cardSize}
        onValueChange={onViewChange}
        className="grid grid-cols-2 gap-2"
      >
        <TooltipProvider>
          <Tooltip>
            <TooltipTrigger asChild>
              <div>
                <RadioGroupItem
                  value="compact"
                  id="compact"
                  className="peer sr-only"
                />
                <Label
                  htmlFor="compact"
                  className={cn(
                    'flex cursor-pointer flex-col items-center justify-between rounded-md border-2 border-muted bg-popover p-4 transition-colors hover:bg-accent hover:text-accent-foreground',
                    cardSize === 'compact' && 'border-primary'
                  )}
                >
                  <div className="grid w-full gap-1 text-center">
                    <div className="space-y-2">
                      <div className="mx-auto h-6 w-8/12 rounded-md bg-muted" />
                    </div>
                  </div>
                  <span className="mt-2 block w-full text-center text-sm font-normal">
                    Compact
                  </span>
                </Label>
              </div>
            </TooltipTrigger>
            <TooltipContent>
              <p>Compact view with title only</p>
            </TooltipContent>
          </Tooltip>

          <Tooltip>
            <TooltipTrigger asChild>
              <div>
                <RadioGroupItem
                  value="default"
                  id="default"
                  className="peer sr-only"
                />
                <Label
                  htmlFor="default"
                  className={cn(
                    'flex cursor-pointer flex-col items-center justify-between rounded-md border-2 border-muted bg-popover p-4 transition-colors hover:bg-accent hover:text-accent-foreground',
                    cardSize === 'default' && 'border-primary'
                  )}
                >
                  <div className="grid w-full gap-1">
                    <div className="space-y-2">
                      <div className="h-2 w-full rounded-md bg-muted" />
                      <div className="h-1 w-full rounded-md bg-muted" />
                    </div>
                  </div>
                  <span className="mt-2 block w-full text-center text-sm font-normal">
                    Default
                  </span>
                </Label>
              </div>
            </TooltipTrigger>
            <TooltipContent>
              <p>Default view with description and tags</p>
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
      </RadioGroup>
    </div>
  )
}
