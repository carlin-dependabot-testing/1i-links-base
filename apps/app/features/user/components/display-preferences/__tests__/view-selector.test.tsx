import { act } from 'react'

import { userEvent } from '@testing-library/user-event'

import { render, screen } from '@/utils/test-utils'

// Import the mocks
import '../../__mocks__/radix-ui'
import { ViewSelector } from '../view-selector'

const renderComponent = (
  props: Partial<Parameters<typeof ViewSelector>[0]> = {}
) => {
  const defaultProps = {
    cardSize: 'default' as const,
    onViewChange: jest.fn()
  }

  return render(<ViewSelector {...defaultProps} {...props} />)
}

describe('ViewSelector', () => {
  it('should render ViewSelector correctly', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('should display both compact and default view options', () => {
    renderComponent()

    expect(screen.getByLabelText('Compact')).toBeInTheDocument()
    expect(screen.getByLabelText('Default')).toBeInTheDocument()
  })

  it('should call onViewChange when selecting a different view', async () => {
    const user = userEvent.setup()
    const onViewChange = jest.fn()

    await act(async () => {
      renderComponent({ onViewChange })
    })

    const compactOption = screen.getByLabelText('Compact')

    await act(async () => {
      await user.click(compactOption)
      // Add a small delay to allow for state updates
      await new Promise(resolve => setTimeout(resolve, 0))
    })

    expect(onViewChange).toHaveBeenCalledWith('compact')
  })

  it('should highlight the selected view option', () => {
    renderComponent({ cardSize: 'compact' })

    const compactLabel = screen.getByText('Compact').parentElement
    const defaultLabel = screen.getByText('Default').parentElement

    expect(compactLabel).toHaveClass('border-primary')
    expect(defaultLabel).not.toHaveClass('border-primary')
  })
})
