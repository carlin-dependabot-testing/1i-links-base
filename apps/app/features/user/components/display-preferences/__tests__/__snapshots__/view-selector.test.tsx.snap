// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewSelector should render ViewSelector correctly 1`] = `
<div>
  <div
    class="group/sidebar-wrapper flex min-h-svh w-full has-[[data-variant=inset]]:bg-sidebar"
    style="--sidebar-width: 16rem; --sidebar-width-icon: 3rem;"
  >
    <script
      nonce=""
    >
      ((e,s,n,m,o,l,u,h)=&gt;{let c=document.documentElement,T=["light","dark"];function p(d){(Array.isArray(e)?e:[e]).forEach(g=&gt;{let k=g==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(c.classList.remove(...S),c.classList.add(d)):c.setAttribute(g,d)}),C(d)}function C(d){h&&T.includes(d)&&(c.style.colorScheme=d)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let d=localStorage.getItem(s)||n,g=u&&d==="system"?a():d;p(g)}catch(d){}})("class","theme","system",null,["light","dark"],null,true,true)
    </script>
    <div
      class="grid gap-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
      >
        View
      </label>
      <div
        data-testid="radio-group"
        data-value="default"
      >
        <div
          data-state="closed"
        >
          <input
            data-testid="radio-item"
            id="compact"
            type="radio"
            value="compact"
          />
          <label
            class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 flex cursor-pointer flex-col items-center justify-between rounded-md border-2 border-muted bg-popover p-4 transition-colors hover:bg-accent hover:text-accent-foreground"
            for="compact"
          >
            <div
              class="grid w-full gap-1 text-center"
            >
              <div
                class="space-y-2"
              >
                <div
                  class="mx-auto h-6 w-8/12 rounded-md bg-muted"
                />
              </div>
            </div>
            <span
              class="mt-2 block w-full text-center text-sm font-normal"
            >
              Compact
            </span>
          </label>
        </div>
        <div
          data-state="closed"
        >
          <input
            data-testid="radio-item"
            id="default"
            type="radio"
            value="default"
          />
          <label
            class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 flex cursor-pointer flex-col items-center justify-between rounded-md border-2 bg-popover p-4 transition-colors hover:bg-accent hover:text-accent-foreground border-primary"
            for="default"
          >
            <div
              class="grid w-full gap-1"
            >
              <div
                class="space-y-2"
              >
                <div
                  class="h-2 w-full rounded-md bg-muted"
                />
                <div
                  class="h-1 w-full rounded-md bg-muted"
                />
              </div>
            </div>
            <span
              class="mt-2 block w-full text-center text-sm font-normal"
            >
              Default
            </span>
          </label>
        </div>
      </div>
    </div>
    <div
      class="fixed bottom-4 right-4 z-50 flex items-center gap-2 rounded-lg bg-background/80 p-2 shadow-sm backdrop-blur-sm"
    />
    <div
      aria-label="Notifications (F8)"
      role="region"
      style="pointer-events: none;"
      tabindex="-1"
    >
      <ol
        class="fixed top-0 z-[100] flex max-h-screen w-full flex-col-reverse p-4 sm:bottom-0 sm:right-0 sm:top-auto sm:flex-col md:max-w-[420px]"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
